Source/Modules/javascript.cxx

```diff
@@ -2852,7 +2847,22 @@ int TypeScriptTypes::memberfunctionHandler(Node *n)
 {
   if (generateTsTypes)
   {
-    tsTypeInterfaceDeclaration->addMemberFunction(n);
+    String *functionName = Getattr(n, "sym:name");
+    String *functionPrototype = NewString(functionName);
+    Append(functionPrototype,"?(");
```

Comment:
It's a bit difficult to see what is this supposed to generate (a comment would probably help), but why should we always have `?` in the prototype?

Reply:
If we don't mark the function as optional, then we cannot assign a variable directly with JSON. Probably this should be controlled by a pragma such as "typescriptoptionalmethods"

Reply:
I'm not sure I understand this. This should be generic code, not related to JSON in any way. Perhaps JSON structure needs to be checked, i.e. it should be explicitly asserted to contain the expected values in our code instead?

Reply:
Sorry, probably I am complicating again :-(. 
Imagine we have following interface definition:

export interface C {
   intValue: number;
   floatValue: number;
   funcWithClass(cVar:C) : A;
}

Due to the function not being optional, this prevents us to do following initialization, Because of the non optional nature of funcWithClass():

const cValue: C = {
    intValue: 1,
    floatValue: 2.3,
};

I use initializations in unit tests. If the functions are not optional, then I cannot use the type when declaring the variable. Probably we wont use that much literal object initializations (sorry: I used the term JSON but the correct is literal object initialization)

Reply:
I'm not a TS expert but it seems strange to make all functions optional to me. And how/what with are they _supposed_ to be initialized anyhow? Something seems to be wrong here, perhaps we need separate types for "in" and "out" interfaces? Because we know that those that we get from native code always have these values. Or maybe we should generate classes implementing these interfaces that can be created from TS side using the appropriate ctors?

Reply:
Your arguments make sense. Let me think a bit more about it.

Reply:
After doing way too many tests and created some examples and explanation, I just realized that I was wanting to do a stupid thing: creating opscore objects from a literal initialization, such as:

let detailsCountry : DetailsCountry = {
  id: "PT",
  name: "Portugal",
  currency: "PTE",
};

instead of instantiating first by using the proper contructor existing in generated c++ wrapper: 

let detailsCountry : DetailsCountry = new datastore.DetailsCountry();
detailsCountry.id = 'PT';
detailsCountry.name = 'Portugal';
detailsCountry.currency = 'PTE';

Literal initialization of objects is pretty common in TypeScript/JavaScript, but now (hope not change again my opinion) I think it should be avoided because we are not creating a propper object that can be used against opscore classes.

So, in conclusion, I will change the function emition to not having optional functions that of course is a anacronic thing.